### PR TITLE
Add remaining json rpc methods and types

### PR DESCRIPTION
**Pull request by knewkirk**
# Add remaining RPC methods

[JSON RPC Spec](https://ethereum.org/en/developers/docs/apis/json-rpc/#json-rpc-methods)
[Wally API Support](https://docs.google.com/spreadsheets/d/137F8veHR96ABgNrBARIGWPoN_gmV12lRTJfRDZke-mw/edit?usp=sharing) _input/feedback welcome!_
[Other Sites' Docs](https://docs.google.com/document/d/1BIVfFSL6tvmWV2FCH1S6dVbi1x3KdhPltcopL81C4N0/edit?usp=share_link)

I've added _(most of)_ the rest of the methods, along with their parameter and return types. There's a bit more type casting and `any` than I'd like, but it was getting pretty hairy and I figured I should stop while I'm ahead*. _(right?)_  

## Overview:
- one main public `request()` function that handles all rpc methods and adheres to the spec.
- a wally-specific function to handle account and signing methods, with parameter and return formatting to work with the api
- an rpc method to handle the rest
- a _whole bunch_ of types, that should hopefully make it easier for folks to develop with/on the sdk.

### Typehinting:
Screen Shot 2022-11-07 at 4 58 56 PM
Screen Shot 2022-11-07 at 4 59 16 PM

### CODE REVIEW

**Review by cjmcassar — APPROVED** [review]

File: src/wally-connector.ts [hunk 1]
```diff
@@ -234,96 +243,168 @@ class WallyConnector {
...
+   * @returns WallyResponse - adheres to the json rpc spec
+   */
+  private async requestWally<T extends WallyMethodName>(
+    method: T,
+    params: WallyMethodParams<T>
+  ): Promise<WallyResponse<T> | null> {
     let resp: Response;
     try {
-      resp = await fetch(`${this.host}/oauth/me`, {
+      resp = await fetch(`${this.host}/oauth${WALLY_ROUTES[method]}`, {
```
Comment (mjmayank) [inline src/wally-connector.ts:333]: is this missing a `/` between oauth and the method route?

File: src/wally-connector.ts [hunk 2]
```diff
@@ -234,96 +243,168 @@ class WallyConnector {
...
+        },
+        body: JSON.stringify({
+          method,
+          params,
+        }),
+      });
+
+      if (!resp.ok || resp.status >= 300) {
+        throw new Error(
+          'Wally server returned a non-successful response when signing a message'
```
Comment (mjmayank) [inline src/wally-connector.ts:390]: does this error message also need to be updated?

File: src/wally-connector.ts [hunk 3]
```diff
@@ -234,96 +243,168 @@ class WallyConnector {
...
 
-    if (!resp.ok || resp.status >= 300) {
-      throw new Error(
-        'Wally server returned a non-successful response when signing a message'
+      const contentType = resp.headers.get('content-type');
+      if (contentType && contentType.indexOf('application/json') !== -1) {
+        const json = await resp.json();
+        return json;
+      } else {
+        const text = await resp.text();
```
Comment (mjmayank) [inline src/wally-connector.ts:399]: when would the response type be text? is this ever expected?

File: src/wally-connector.ts (hunk 2 above)
Comment (knewkirk) [reply src/wally-connector.ts:390]: good catch!

File: src/wally-connector.ts (hunk 1 above)
Comment (knewkirk) [reply src/wally-connector.ts:333]: no, but it doesn't look right, huh? 😅  i'll move the leading slashes out of the constants to here instead ⚡

File: src/wally-connector.ts [hunk 4]
```diff
@@ -234,96 +243,168 @@ class WallyConnector {
...
-    switch (req.method) {
-      case MethodName.REQUEST_ACCOUNTS:
-        return this.requestAccounts();
-      // TODO: figure out which name to use
-      case MethodName.PERSONAL_SIGN:
-      case MethodName.SIGN:
-        return this.signMessage(req.params);
-      case MethodName.GET_BALANCE:
-        return this._request(req.method, req.params);
+    if (Object.values(WallyMethodName).indexOf(req.method as any) > -1) {
```
Comment (mjmayank) [inline src/wally-connector.ts:269]: could we make this into a helper function just to make the code easier to read?

File: src/wally-connector.ts [hunk 5]
```diff
@@ -234,96 +243,168 @@ class WallyConnector {
...
+      return this.requestWally(
+        req.method as WallyMethodName,
+        'params' in req ? (req.params as WallyMethodParams<T>) : undefined
+      ) as Promise<WallyResponse<T>>;
+    } else if (Object.values(RPCMethodName).indexOf(req.method as any) > -1) {
```
Comment (mjmayank) [inline src/wally-connector.ts:274]: same here

File: src/wally-connector.ts [hunk 6]
```diff
@@ -234,96 +243,168 @@ class WallyConnector {
...
+    } else {
+      throw new Error(`Method ${req.method} is unsupported at this time.`);
+    }
+  }
+
+  private formatWallyParams<T extends WallyMethodName>(
+    method: T,
+    params: WallyMethodParams<T>
+  ): string {
+    if (method === WallyMethodName.SIGN) {
```
Comment (mjmayank) [inline src/wally-connector.ts:288]: should this also be a case statement just for consistency?

File: src/constants.ts [hunk 7]
```diff
@@ -57,4 +59,14 @@ export const getRedirectPage = (): HTMLElement => {
...
   return containerEl;
-};
\ No newline at end of file
+};
+
+export const WALLY_ROUTES: Record<WallyMethodName, string> = {
+  [WallyMethodName.ACCOUNTS]: 'me',
+  [WallyMethodName.REQUEST_ACCOUNTS]: 'me',
+  [WallyMethodName.SIGN]: 'wallet/sign-message',
+  [WallyMethodName.PERSONAL_SIGN]: 'wallet/sign-message',
```
Comment (mjmayank) [inline src/constants.ts:68]: some of these endpoints still need to be written right?

Comment (knewkirk) [reply src/constants.ts:68]: nah man we're all set 🐴 
https://github.com/Wally-xyz/wallet-api/pull/221

File: src/wally-connector.ts (hunk 4 above)
Comment (knewkirk) [reply src/wally-connector.ts:269]: for sure 👍  - funny you mention it, I was actually trying to get a type guard function setup here earlier and gave up haha - i'll get it figured

File: src/constants.ts (hunk 7 above)
Comment (knewkirk) [reply src/constants.ts:68]: unless you're specifically talking about `eth_sign` vs `personal_sign` - still not sure about the crypographic differences there - they just go to that same api endpoint, i'll add a note.

Comment (mjmayank) [reply src/constants.ts:72]: nah i just meant do they exist at all. cool, forgot we added that

**Review by mjmayank — APPROVED** [review]

**Pull request merged**